### ISSUE TITLE
v3: Invalid output (bad indentation) when an indentation indicator/hint is used in a scalar block

### ISSUE THREAD

**Issue by abursavich** [go-yaml/yaml#643]
The indentation change in https://github.com/go-yaml/yaml/commit/a5ece683394c3b88d90572e44d36c93aea492c2c is causing  invalid output when an indentation indicator/hint is used in a scalar block: https://play.golang.org/p/tbWDmsaqANV

for:
```go
yaml.Marshal([]string{" hello\nworld"})
```
expecting:
```yaml
- |4-
     hello
    world
```
got:
```yaml
- |4-
   hello
  world
```
which is invalid and cannot be parsed.

This relates to #633 and I had commented about it there, but invalid output is more severe than unpleasant indentation... so I figured it deserved a separate issue.

**Comment by niemeyer** [comment]
Thanks for the report and PR. I'll look into those.

Meanwhile, have you signed the contributor agreement (https://ubuntu.com/legal/contributors)?

It's easy to sign online and the text should be very reasonable hopefully.

**Comment by abursavich** [comment]
Yeah, I signed it the other day and it required a Canonical manager/contact so I put "niemeyer" for that? Wasn't sure what it wanted there.

**Comment by kushagharahi** [comment]
^Apologies about the spam above.
I'm somewhat stuck between a rock and a hard place.
In V2 I'm running into this issue -> https://github.com/go-yaml/yaml/issues/591 which is fixed in v3, in return in v3 I'm running into this issue above.

It would be awesome if the above PR was merged. Happy to help make any changes that are necessary.

**Comment by ccampo133** [comment]
Any chance we could get this merged? PR looks good to me but obviously I am not an approver ;)

**Comment by hedhyw** [comment]
```go
enc := yaml.NewEncoder(w)
enc.SetIndent(2)
```

Changing indent to 2 can solve the issue before https://github.com/go-yaml/yaml/pull/864 is merged.

https://github.com/grpc-ecosystem/grpc-gateway/pull/2801/files#diff-7afc99d2a20e8d539a26ebfcdd433b5273995cf9021972f2d0ef93a5848e0280R35

**Comment by emanlodovice** [comment]
Hi @niemeyer , thank you for your work on this package. Is there any chance we could have the PR merged? Thanks.

**Comment by lucastheisen** [comment]
This is likely related but a [second repro](https://go.dev/play/p/8lNQjhHKiH8):

```go
package main

import (
	"fmt"

	"gopkg.in/yaml.v3"
)

func main() {
	var data map[any]any
	err := yaml.Unmarshal([]byte(`---
foo:
  bar: |4
     `+"```"+`bash
    ps -aef
    `+"```"+`
`), &data)

	fmt.Printf("err=[%v]\n", err)
	fmt.Printf("foo=[%v]\n", data)
}
```

Results in:

```console
err=[yaml: line 4: found character that cannot start any token]
foo=[map[]]
```

**Comment by lucastheisen** [comment]
[this playground](https://go.dev/play/p/q4IjHk7oP-x) shows another repro of this issue in a round trip (Marshal -> Unmarshal):

```go
package main

import (
	"fmt"

	"github.com/sashabaranov/go-openai"
	"gopkg.in/yaml.v3"
)

func main() {
	data := openai.ChatCompletionRequest{
		Model: "mistral",
		Messages: []openai.ChatCompletionMessage{
			{
				Role: "assistant",
				Content: ` ` + "```" + `bash

        #!/bin/bash

        # Function to read null-terminated list from standard input
        read_null_term_list() {
         local IFS='' read -r -d '' line
         local list=()
         while IFS= read -r line; do
          list+=("$line")
         done
         eval "$@ \${list[@]}"
        }
` + "```" + `
`,
			},
		},
	}

	yml, err := yaml.Marshal(data)
	fmt.Printf("err=[%v]\n", err)
	fmt.Printf("yml=[\n%s\n]\n", yml)

	var data2 openai.ChatCompletionRequest
	err = yaml.Unmarshal(yml, &data2)
	fmt.Printf("err=[%v]\n", err)
	fmt.Printf("unmarshaled=[%v]\n", data2)
}

```

The suggested [`SetIndent(2)` workaround](https://github.com/go-yaml/yaml/issues/643#issuecomment-1189154378) seems to [work](https://go.dev/play/p/ido343N9gbG).
